pkg/image/cdi/download.go

```diff
@@ -1,19 +1,299 @@
+	contentType := downloadResp.Header.Get("Content-Type")
+	if contentType != "" {
+		rw.Header().Set("Content-Type", contentType)
+	}
+
+	if _, err := io.Copy(rw, downloadResp.Body); err != nil {
+		return fmt.Errorf("failed to copy download content to target(%s), err: %w", targetFileName, err)
+	}
+
 	return nil
 }
+
+func (cd *Downloader) generateDownlodServerPod(podName, vmImageName string) *corev1.Pod {
+	virtImageStr := cd.getVirtHandlerImage()
+	clusterRepoImageStr := cd.getClusterRepoImage()
+
+	convertCmd := fmt.Sprintf("qemu-img convert -f raw /tmp/image-vol -O qcow2 -c -S 4K /image-dir/%s.qcow2", vmImageName)
+
+	targetPod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: podName,
+			Labels: map[string]string{
+				"app": podName,
+			},
+		},
+		Spec: corev1.PodSpec{
+			Volumes: []corev1.Volume{
+				{
+					Name: "image-vol",
+					VolumeSource: corev1.VolumeSource{
+						PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+							ClaimName: vmImageName,
+						},
+					},
+				},
+				{
+					Name: "image-dir",
+					VolumeSource: corev1.VolumeSource{
+						EmptyDir: &corev1.EmptyDirVolumeSource{},
+					},
+				},
+			},
+			InitContainers: []corev1.Container{
+				{
+					Name:            "image-coverter",
+					Image:           virtImageStr,
+					ImagePullPolicy: corev1.PullIfNotPresent,
+					VolumeDevices: []corev1.VolumeDevice{
+						{
+							Name:       "image-vol",
+							DevicePath: "/tmp/image-vol",
+						},
+					},
+					VolumeMounts: []corev1.VolumeMount{
+						{
+							Name:      "image-dir",
+							MountPath: "/image-dir",
+						},
+					},
+					Command: []string{"/bin/sh", "-c"},
+					Args: []string{
+						convertCmd,
+					},
+				},
+			},
+			Containers: []corev1.Container{
+				{
+					Name:            "image-exporter",
+					Image:           clusterRepoImageStr,
+					ImagePullPolicy: corev1.PullIfNotPresent,
+					Ports: []corev1.ContainerPort{
+						{
+							ContainerPort: 80,
+							Protocol:      corev1.ProtocolTCP,
+						},
+					},
+					VolumeMounts: []corev1.VolumeMount{
+						{
+							Name:      "image-dir",
+							MountPath: "/srv/www/htdocs/images/",
+						},
+					},
+					Command: []string{"/bin/sh", "-c"},
+					Args: []string{
+						"nginx -g 'daemon off;' && while true; do sleep 3600; done",
+					},
+				},
+			},
+		},
+	}
+
+	return targetPod
+}
+
+func (cd *Downloader) getVirtHandlerImage() string {
+	image, err := utilHelm.FetchImageFromHelmValues(cd.clientSet,
+		util.HarvesterSystemNamespaceName,
+		util.HarvesterChartReleaseName,
+		[]string{"kubevirt-operator", "containers", "handler", "image"})
+	if err != nil {
+		return ""
+	}
+	targetImage := fmt.Sprintf("%s:%s", image.Repository, image.Tag)
+	return targetImage
+}
+
+func (cd *Downloader) getClusterRepoImage() string {
+	deployContent, err := cd.clientSet.AppsV1().Deployments("cattle-system").Get(context.TODO(), "harvester-cluster-repo", metav1.GetOptions{})
+	if err != nil {
```

Comment:
we absorb the error and return "" when lookup fails for any reason.

the pod creation will likely fail because container has no image name specified. are we ok with this? I do believe this is an edge case since the repo should be running before harvester initialises

Reply:
With the new mechanism for the downloader, the wrong repo name will not make the corresponding CRD become ready. So we will know what happened.